---
title: "MicrobiomeWorkshopII.Rmd"
abstract:>
The second part of the workshop  demonstrates how to use `dada2` on raw reads,
and analysis of these data using the `phyloseq`, `treeDA`, `adaptiveGPCA` packages for denoising, estimating differential abundance, ordinations.
The `treelapse` and  `metavizr` packages allow browsing and interactive visualization of microbiome profiles. Together, these packages provide easily linked components for data acquisition and flexible analysis of 16S rRNA and whole metagenome shotgun microbiome profiles. At the end of this workshop, users will be able to access publicly available metagenomic data and to perform common statistical analyses of these and other data in Bioconductor.
author: "Susan Holmes"
date: "July 17, 2017"
output: 
   BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{MicrobiomeWorkshopII}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r NeededPackages}
cranpkgs=c("ggplot2","devtools","adaptivGPCA","treelapse", "structSSI", "vegan", "phangorn")
BioCpkgs=c("dada2", "curatedMetagenomicData", "phyloseq", "DECIPHER")
ghpkgs= c("treeDA", "treelapse","metavizr")
instp <-  cranpkgs %in% installed.packages()
if(any(!instp)) {
   install.packages(cranpkgs[!instp],repos="https://cloud.r-project.org")
}
instp <- BioCpkgs %in% installed.packages()
if(any(!instp)) {
   source("http://bioconductor.org/biocLite.R")
   biocLite(BioCpkgs [!instp], ask = F)
}

## Check for github packages
#instp <- ghpkgs %in% installed.packages()
if(!("treeDA"  %in% installed.packages()))
   devtools::install_github("jfukuyama/treeDA", dependencies = TRUE)
if(!("treelapse"  %in% installed.packages()))
  devtools::install_github("krisrs1128/treelapse", dependencies = TRUE)
if(!("metavizr"  %in% installed.packages()))
devtools::install_github("epiviz/metavizr", build_vignettes = TRUE)

sapply(c(cranpkgs, BioCpkgs, ghpkgs), require, character.only = TRUE)

```



# dada2 theory

CLustering using frequencies not just distances.

# dada2 tutorial

[See tutorial here](https://benjjneb.github.io/dada2/tutorial.html)


## Components of a phyloseq object

This  `r BiocStyle::Biocpkg("phyloseq")` objects contain 3 components, with extractor functions hinted at by its show method:

```{r}
loman.pseq
```

`otu_table()` returns the same thing as `exprs( loman.eset )` did, the Operational Taxanomic Unit (OTU) table. Here are the first 6 rows and 5 columns:

```{r, warning=FALSE}
otu_table( loman.pseq )[1:6, 1:5]
```

The same patient or participant data that was available from  `pData( loman.eset )` is now availble using `sample_data()` on 
the `r BiocStyle::Biocpkg("phyloseq")` object:

```{r, warning=FALSE}
sample_data( loman.pseq )[1:6, 1:5]
```

But this object also is aware of the taxonomic structure, which will enable the powerful subsetting methods of the `r BiocStyle::Biocpkg("phyloseq")` package.

```{r, warning=FALSE}
head( tax_table( loman.pseq ) )
```

## Subsetting / Pruning

The process of subsetting begins with the names of phylogenetic ranks:

```{r, warning=FALSE}
rank_names( loman.pseq )
```

Taxa can be filtered by these rank names. For example, to return an object with only species and strains:

```{r, warning=FALSE}
subset_taxa( loman.pseq, !is.na(Species))
```

To keep only phylum-level data (not class or finer, and not kingdom-level):

```{r}
subset_taxa( loman.pseq, is.na(Class) & !is.na(Phylum))
```

Or to keep only Bacteroidetes phylum. Note that taxa names have been shortened from the rownames of the `ExpressionSet` object, for nicer plotting.

```{r}
loman.bd = subset_taxa( loman.pseq, Phylum == "Bacteroidetes")
head( taxa_names( loman.bd ) )
```

## Advanced Pruning

The `r BiocStyle::Biocpkg("phyloseq")` package provides advanced pruning of taxa, such as the following which keeps only taxa that are among the most abundant 5% in at least five samples:

```{r, warning=FALSE}
keepotu = genefilter_sample(loman.pseq, filterfun_sample(topp(0.05)), A=5)
summary(keepotu)
subset_taxa(loman.pseq, keepotu)
```

Note that `r BiocStyle::Biocpkg("phyloseq")` also provides `topk()` for selecting the most abundant `k` taxa, and other functions for advanced pruning of taxa.

## Taxonomy Heatmap

The `r BiocStyle::Biocpkg("phyloseq")` package provides the `plot_heatmap()` function to create heatmaps using a variety of built-in dissimilarity metrics for clustering. Here, we apply the same abundance filter as above, keep only strain-level OTUs. This function supports a large number of distance and ordination methods, here we use Bray-Curtis dissimilarity for distance and PCoA as the ordination method for organizing the heatmap.

```{r, warning=FALSE}
loman.filt = subset_taxa(loman.pseq, keepotu & !is.na(Strain))
plot_heatmap(loman.filt, method="PCoA", distance="bray")
```

## Taxonomy Histogram

Here we plot the top 20 most abundant species (not strains), defined by the sum of abundance across all samples in the dataset:

```{r, warning=FALSE}
loman.sp = subset_taxa(loman.pseq, !is.na(Species) & is.na(Strain))
par(mar = c(20, 4, 0, 0) + 0.15) #increase margin size on the bottom
barplot(sort(taxa_sums(loman.sp), TRUE)[1:20] / nsamples(loman.sp),
        ylab = "Total counts", las = 2)
```

## Alpha Diversity Estimation

The `r BiocStyle::Biocpkg("phyloseq")` package calculates numerous alpha diversity measures. Here we compare three diversity in the species-level data, stratifying by stool texture:

```{r, warning=FALSE}
alphas = c("Shannon", "Simpson", "InvSimpson")
plot_richness(loman.sp, "stool_texture", measures = alphas)
```
Let's compare these three alpha diversity measures:

```{r, warning=FALSE}
pairs( estimate_richness(loman.sp, measures = alphas) )
```

## Beta Diversity / Dissimilarity Clustering

Numerous beta diversity / dissimilarity are provided by the `distance()` function when provided a `r BiocStyle::Biocpkg("phyloseq")` object, and these can be used for any kind of clustering or classification scheme. 

For example, here is a hierarchical clustering dendrogram produced by the `hclust()` from the base R `stats` package with "Ward" linkage:

Fixme:
Probably want to change this to use of pheatmap or neatmap.
Poor quality output.
```{r, warning=FALSE}
mydist = distance(loman.sp, method="bray")
myhclust = hclust( mydist )
plot(myhclust, main="Bray-Curtis Dissimilarity", 
     method="ward.D", xlab="Samples", sub = "")
```


## Ordination Analysis

The `r BiocStyle::Biocpkg("phyloseq")` package provides a variety of ordination methods, with convenient options for labelling points. Here is a 
Principal Coordinates Analysis plot of species-level taxa from the Loman dataset, using Bray-Curtis distance:

```{r, warning=FALSE}
ordinated_taxa = ordinate(loman.sp, method="PCoA", distance="bray")
plot_ordination(loman.sp, ordinated_taxa, color="stool_texture", 
                title = "Bray-Curtis Principal Coordinates Analysis")
```


```{r}
plot_scree(ordinated_taxa, title="Screeplot")
```

# DPCOA and adaptive GPCA analyses

It is important to be able to integrate the phylogenetic tree information into
the analysis. 
We start with a small example from the Antibiotic data study (Relman and Dethfelsen, 2011)
available in the DPCoA packages.
This can be done through either the use of the ordinate
function with the method="DPCoA" or better still by using the new adaptiveGPCA
package:

```{r}
data(AntibioticSmall)
out.ff = gpcaFullFamily(AntibioticSmall$X, AntibioticSmall$Q, k = 2)
```

Interpolation between complete tree based DPCOA analyses and
PCA analysis, this requires interactive use and shiny, so not evaluated
in vignette here:

```{r, eval =FALSE}
#not run
data(AntibioticPhyloseq)
pp = processPhyloseq(AntibioticPhyloseq)
out.ff = gpcaFullFamily(pp$X, Q = pp$Q, D = pp$D, k = 2)
out.agpca = visualizeFullFamily(out.ff,
    sample_data = sample_data(AntibioticPhyloseq),
    sample_mapping = aes(x = Axis1, y = Axis2, color = condition),
    var_data = tax_table(AntibioticPhyloseq),
    var_mapping = aes(x = Axis1, y = Axis2, color = Phylum))
```


# Sparse discriminant analyses using a phylogenetic tree

The package itself needs to downloaded from github:

```{r}
##devtools::install.github("jfukuyama/treeDA")
library(treeDA)
library(adaptiveGPCA)
data(AntibioticPhyloseq)
theme_set(theme_bw())
```

```{r}
out.treeda = treeda(response = sample_data(AntibioticPhyloseq)$type,
    predictors = otu_table(AntibioticPhyloseq),
    tree = phy_tree(AntibioticPhyloseq), p = 15)
```

# Multi-table analyses

##Sparse canonical correlations

# Looking at longitudinal data in the presence of a tree using treelapse.

[Tutorial to follow](http://statweb.stanford.edu/~kriss1/antibiotic.html)

```{r}
library(treelapse)
```



TODO:
# Other visualizations using metavizr


